src/annmas/segment/command.py

```diff
@@ -47,60 +84,334 @@ def main(model, threads, output_bam, input_bam):
+                pg_dict = {
+                    "ID": "annmas-segment-0.0.1",
+                    "PN": "annmas",
+                    "VN": "0.0.1",
+                    "DS": "Apply annotation and segmentation model to BAM file.",
+                    "CL": " ".join(sys.argv),
+                }
+                if "PG" in out_bam_header_dict:
+                    out_bam_header_dict["PG"].append(pg_dict)
+                else:
+                    out_bam_header_dict["PG"] = [pg_dict]
+
+                if split_bam:
+                    split_bam_out = pysam.AlignmentFile(
+                        split_bam, "wb", header=out_bam_header_dict
+                    )
+
+                with pysam.AlignmentFile(
+                    output_bam, "wb", header=out_bam_header_dict
+                ) as out_bam_file:
+
+                    for future in concurrent.futures.as_completed(
+                        future_to_segmented_read
+                    ):
+                        read = future_to_segmented_read[future]
+                        try:
+                            path, logp = future.result()
+                        except Exception as ex:
+                            logger.error("%r generated an exception: %s", read, ex)
+                        else:
+                            # Condense the output annotations so we can write them out with indices:
+                            segments = _collapse_annotations(path)
+
+                            # Obligatory log message:
+                            logger.debug(
+                                "Path for read %s (%2.2f): %s",
+                                read.query_name,
+                                logp,
+                                segments,
+                            )
+
+                            # Set our tag and write out the read to the annotated file:
+                            read.set_tag("SG", "|".join([s.to_tag() for s in segments]))
+                            out_bam_file.write(read)
+
+                            # Write out our segmented reads if we want them:
+                            if split_bam_out:
+                                num_segments += _write_segmented_read(
+                                    read, segments, do_simple_splitting, split_bam_out
+                                )
+                            # Increment our counters:
+                            num_reads_annotated += 1
+                            num_sections += len(segments)
+
+                            pbar.update(1)
+
+        logger.info("annmas: segment finished.")
+        logger.info(
+            f"annmas: segmented {num_reads_annotated} reads into {num_sections} sections."
+        )
+
+    finally:
+        if split_bam_out:
+            split_bam_out.close()
+            logger.info(f"annmas: wrote {num_segments} segments.")
+
+
+def _write_segmented_read(read, segments, do_simple_splitting, bam_out):
```

Comment:
I think I'm not quite understanding some of the considerations being made here.  Let's go over this sometime this week.